components/zip321-parse/src/lib.rs

```diff
@@ -0,0 +1,904 @@
+    /// by comparing a two transaction requests for equality after normalization.
+    #[cfg(any(test, feature = "test-dependencies"))]
+    pub fn normalize_and_eq(&mut self, b: &mut Self) -> bool
+    where
+        P: PartialEq,
+    {
+        self.normalize();
+        b.normalize();
+
+        self == b
+    }
+
+    /// Convert this request to a URI string.
+    ///
+    /// Returns None if the payment request is empty.
+    pub fn to_uri(&self) -> String {
+        fn payment_params(
+            payment: &Payment,
+            payment_index: Option<usize>,
+        ) -> impl IntoIterator<Item = String> + '_ {
+            std::iter::empty()
+                .chain(Some(render::amount_coins_zats_param(
+                    (payment.amount_coins, payment.amount_fractional_zatoshis),
+                    payment_index,
+                )))
+                .chain(
+                    payment
+                        .memo
+                        .as_ref()
+                        .map(|m| render::memo_param(&m.to_base64(), payment_index)),
+                )
+                .chain(
+                    payment
+                        .label
+                        .as_ref()
+                        .map(|m| render::str_param("label", m, payment_index)),
+                )
+                .chain(
+                    payment
+                        .message
+                        .as_ref()
+                        .map(|m| render::str_param("message", m, payment_index)),
+                )
+                .chain(
+                    payment
+                        .other_params
+                        .iter()
+                        .map(move |(name, value)| render::str_param(name, value, payment_index)),
+                )
+        }
+
+        match self.payments.len() {
+            0 => "zcash:".to_string(),
+            1 if *self.payments.iter().next().unwrap().0 == 0 => {
+                let (_, payment) = self.payments.iter().next().unwrap();
+                let payment = payment.as_ref();
+                let query_params = payment_params(payment, None)
+                    .into_iter()
+                    .collect::<Vec<String>>();
+
+                format!(
+                    "zcash:{}{}{}",
+                    payment.recipient_address,
+                    if query_params.is_empty() { "" } else { "?" },
+                    query_params.join("&")
+                )
+            }
+            _ => {
+                let query_params = self
+                    .payments
+                    .iter()
+                    .flat_map(|(i, payment)| {
+                        let idx = if *i == 0 { None } else { Some(*i) };
+                        let primary_address = payment.as_ref().recipient_address.clone();
+                        std::iter::empty()
+                            // TODO(schell): mention changed
+                            .chain(Some(render::addr_param_encoded(&primary_address, idx)))
```

Comment:
This changed when it was moved to _not_ encode the address, as a `zip321_parse::Payment`'s address is already encoded (it doesn't get decoded and validated until the "last mile").